tga-core/src/main/kotlin/org/plan/research/tga/core/tool/TestSuite.kt

```diff
@@ -10,5 +10,7 @@ data class TestSuite(
     @Serializable(with = PathAsStringSerializer::class)
     val testSrcPath: Path,
     val tests: List<String>,
+    val testCasesOnly: List<String>,
```

Comment:
Can you explain what is the difference between `tests` and `testCasesOnly` in this context? Why do we need them both?

And what is the purpose of `testSuiteQualifiedName`?

Reply:
1. `tests` - all compilable test cases and the test suite (i.e. `[package].GeneratedTest`).
2. `testCasesOnly` - all compilable test cases without the test suite.
3. `testSuiteQualifiedName` - the test suite in the form `[package].GeneratedTest`.

**Here is an example:**
I have generated tests for the project `traccar-413d9a49c4` and  TestSpark generated the following test cases (and saved them into separate files):
1. `GeneratedDecodeAlarmTest.java` (non compilable).
2. `GeneratedDecodeFaultCodesTest.java` (compilable).
3. `GeneratedDecodeHsoReqTest.java` (compilable).
4. `GeneratedDecodeLoginTest.java` (compilable).
5. `GeneratedSendResponseTest` (non compilable).
6. And the test suite stored in `GeneratedTest.java` (compilable) containing the compilable test cases:
    1. `decodeLoginTest` (from `GeneratedDecodeLoginTest.java`)
    2. `decodeHsoReqTest` (from `GeneratedDecodeHsoReqTest.java`)
    3. `decodeFaultCodesTest` (from `GeneratedDecodeFaultCodesTest.java`)
    
In the above configuration the `testSuite.json` will contain:
```json
    "tests": [
        "org.traccar.protocol.GeneratedDecodeLoginTest",
        "org.traccar.protocol.GeneratedDecodeHsoReqTest",
        "org.traccar.protocol.GeneratedTest",
        "org.traccar.protocol.GeneratedDecodeFaultCodesTest"
    ],
    "testCasesOnly": [
        "org.traccar.protocol.GeneratedDecodeLoginTest",
        "org.traccar.protocol.GeneratedDecodeHsoReqTest",
        "org.traccar.protocol.GeneratedDecodeFaultCodesTest"
    ],
    "testSuiteQualifiedName": "org.traccar.protocol.GeneratedTest",
```
I use there 3 sets of compilable test cases in order to compute coverage robustly: if `tests` are not compilable together, than I try to compile `testCasesOnly` and compute coverage on them, if `testCasesOnly` are also uncompilable I try to compile `testSuiteQualifiedName` (i.e. the test suite alone) and compute coverage on it. 

All of these 3 sets of tests are identical in terms of producible coverage but in some cases some of these sets may not be compilable, (e.g., there exist scenarios when compiling test cases and the test suite together leads to an error (e.g., duplicate classes), although test cases and the test suite, if separated, are compilable).

Likely, `"tests"` may be removed, and the coverage may be computed only on either `testCasesOnly` or `testSuiteQualifiedName` but the `tests` prop is used throughout the pipeline, so I decided not to modify/remove this prop.

Reply:
Ok, got it. But doesn't it mean that `"org.traccar.protocol.GeneratedTest"` should always contain something that can be considered a 'compilable test suite produced by TestSpark'? I just don't really like the idea of extending the common API of the pipeline with TestSpark specific properties

Reply:
> Ok, got it. But doesn't it mean that `"org.traccar.protocol.GeneratedTest"` should always contain something that can be considered a 'compilable test suite produced by TestSpark'?

Yes, this is true but there is a bug in TestSpark which may render the test suite non-compilable: The set of imports that TestSpark saves may not be sufficient (see [issue-176](https://github.com/JetBrains-Research/TestSpark/issues/176)). Thus, the test suite has totally valid test cases but some imports may be missing, because of this I cannot rely solely on `testSuiteQualifiedName`. 

In order to fix the issue above the imports merger should be implemented. 

I am running very short on the time left before the intermediate thesis submission deadline, thus I have no time to focus on fixing that issue and testing the solution. I can get to it after 21st of April.

Reply:
Ok, I'll merge this version then and we will think about refactoring it later